### ISSUE TITLE
Insane(100%) consumption of resources

### ISSUE THREAD

**Issue by sudeepgaddam** [bazelbuild/bazel#1160]
I tried to build tensorflow source using bazel. I was assuming bazel would take care of parallelisation in a smart way by having a balance between performance and judicious usage. But it was using 100% of all the cores and 95% RAM. This is in no way a good approach. I had to hard boot the  system.

Under default cases, Bazel should use resources in more smarter way.

**Comment by trainman419** [comment]
I've seen the same issue on my machine, building a large custom codebase.

The only solution I've found is to set the --jobs option to bazel to limit the number of parallel builds that it does.

Since my machine uses about 50% RAM under nomal operating conditions, and I wanted to leave some headroom, I tried setting `--ram_utilization_factor` to a lower number (20 or 30), but doing a build still results in 100% memory usage and swapping.

From reading the documentation, it seems like bazel should be estimating the memory usage of each build command and attempting to keep the overall memory usage below this value, but this doesn't appear to work for me.

This has also shown up on stack overflow, with similar results: http://stackoverflow.com/questions/34382360/decrease-bazel-memory-usage/34399184#34399184

If it helps, the build that I'm doing is primarily large, heavily templated C++ code, where a single instance of the compiler or the linker can use multiple GB of RAM, on a machine that has 16GB of RAM.

**Comment by damienmg** [comment]
Yes sorry for that the ressource estimation in Bazel is far from optimal and I know @philwo has been working on improving it but it's tuning some parameters. In the time being the workaround is indeed to manually set `--jobs` to limit the number of parallel jobs bazel runs.

**Comment by philwo** [comment]
While we have a plan for a simplified resource management system, I'm not sure if that would fix this case. The problem is that we can't reliably estimate in advance how much RAM a given action needs and also AFAIK Bazel bases its "RAM utilization factor" on the assumption that it is the only process on the system wanting to use RAM - so "80%" means "Bazel can use up to 80% of the total installed RAM" not "Bazel should make sure that the system always keeps 20% RAM free".

**Comment by 3ygun** [comment]
@damienmg is there any information I can provide you with to help fix this error? I've tried to build TensorFlow with Bazel and seem to be running into this error. I've been able to repeated a total system freeze after presumably running out of RAM 4 times now. Will update this post if the `--jobs` argument helps.

Type | Info
--- | ---
System Information | Lenovo y510p running Fedora 25 (8GB RAM, 8 cores)
Bazel | Version 0.4.5- (@non-git) installed from source
Using with | open JDK 1.8.0.121
Building command | `bazel build -c opt //tensorflow/examples/android:tensorflow_demo`

Below is an htop when my system froze (idle RAM is ~1GB):
![img_20170411_150812](https://cloud.githubusercontent.com/assets/5482320/24926619/429297c6-1eca-11e7-99ae-2a4326381ee9.jpg)

**Comment by philwo** [comment]
@3ygun Try specifying --jobs=4 - that way, Bazel should never spawn more than 4 processes in parallel. If it works fine for your system, you could put it into your ~/.bazelrc like this:

```
build --jobs=4
```

**Comment by 3ygun** [comment]
@philwo Thanks! I'm trying `--jobs=20` and getting different build errors but no system freezes so 🤞 and progress?

**Comment by philwo** [comment]
Hi @3ygun,

would you mind trying the [ps_mem.py](https://github.com/pixelb/ps_mem) script to get an accurate memory consumption snapshot when your build hits e.g. 6 GB of RAM? It should print a list of processes sorted by their RAM consumption correctly adjusted for shared memory pages.

The output might help to figure out how we can improve the resource handling for your use case. If it's just one kind of action / program that consistently uses so much RAM, we might add a way to limit the parallelism of actions per mnemonic.

Cheers,
Philipp

**Comment by 3ygun** [comment]
@philwo Will do when I get a chance!

**Comment by 3ygun** [comment]
@philwo sorry for the delayed response but I haven't/don't know how to give you good data and believe the root of my "freezing" problem comes from some bugs in Fedora 25. This resulted from the problem coming up all to frequently and having had it completely freeze with Firefox and a terminal open...

Alas using `--jobs=1` has allowed my builds of TensorFlow to complete well multiple times. Sorry, I cannot help more.

I could try using a different command from the linked `ps_mem.py`. In the past I've just tried `sudo ps_mem` which just lists all current process RAM amounts so not useful. Is there anything specific I should try?
